### PR TITLE
8304888: Add dedicated VMProps for linker and fallback linker

### PR DESCRIPTION
**Pull request by JornVernee**
This patch adds a dedicated jtreg property to check which CABI is being used by the JDK, which can be used both to check whether the foreign linker is supported, and whether the fallback linker is being used. (and, possibly it can also be use to check for a particular ABI in case we want to add ABI specific tests).

Checking whether the foreign linker is supported currently requires using an unwieldy expression that checks if we are running on a platform that has a foreign linker port. Checking for the fallback linker currently uses `vm.flavor != "zero"` which is not always correct, since the fallback linker can also be used on other platforms which are not `zero`.

To initialize the property, VMProps directly reads `jdk.internal.foreig.CABI::current()`. Since this class is in an internal package, `--add-exports` flags are added as javac flags and VM flags for the extra prop definitions class.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304888](https://bugs.openjdk.org/browse/JDK-8304888): Add dedicated VMProps for linker and fallback linker

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [272478b6](https://git.openjdk.org/jdk/pull/13429/files/272478b6e9f4b892b775fda5a7b00702e5156bcb)
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [272478b6](https://git.openjdk.org/jdk/pull/13429/files/272478b6e9f4b892b775fda5a7b00702e5156bcb)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13429/head:pull/13429` \
`$ git checkout pull/13429`

Update a local copy of the PR: \
`$ git checkout pull/13429` \
`$ git pull https://git.openjdk.org/jdk.git pull/13429/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13429`

View PR using the GUI difftool: \
`$ git pr show -t 13429`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13429.diff">https://git.openjdk.org/jdk/pull/13429.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13429#issuecomment-1519629282)

### CODE REVIEW

File: test/jtreg-ext/requires/VMProps.java [hunk 1]
```diff
@@ -652,6 +655,10 @@ private String isFlagless() {
         return "" + result;
     }
```
Comment (AlanBateman) [inline test/jtreg-ext/requires/VMProps.java:657]: It might be useful to add a comment on `cabi()` as it may not be obvious that this is C ABI.

**Review by AlanBateman — APPROVED** [review]

**Review by AlanBateman — APPROVED** [review]
I looked at the update (comment on VMProps. cabi, update to TestLinker, and s/UNKNOWN/UnSUPPORTED). Looks okay to me.

File: test/jdk/java/foreign/SafeFunctionAccessTest.java [hunk 2]
```diff
@@ -24,7 +24,7 @@
 /*
  * @test
  * @enablePreview
- * @requires ((os.arch == "amd64" | os.arch == "x86_64") & sun.arch.data.model == "64") | os.arch == "aarch64" | os.arch == "riscv64"
+ * @requires jdk.CABI != "UNSUPPORTED"
```
Comment (mcimadamore) [inline test/jdk/java/foreign/SafeFunctionAccessTest.java:27]: what about the name `jdk.native.abi` ?

**Review by mcimadamore — APPROVED** [review]
Looks good - I'm not too much of a fan of "CABI" as a property name - but that's a minor nit. I think the improvements in the jtreg test headers are very nice.

File: test/jdk/java/foreign/SafeFunctionAccessTest.java (hunk 2 above)
Comment (JornVernee) [reply test/jdk/java/foreign/SafeFunctionAccessTest.java:27]: Not so sure. The CABI is really the linker kind, not the underlying ABI ('CABI' is probably a misnomer at this point).

If CABI is bad, I think `jdk(.foreign).linker` would be better as a property name

**Comment by mcimadamore** [comment]
> I went with the `jdk.CABI -> jdk.foreign.linker` name change. Re-tested. Will integrate soon.

Looks good (if we wanted to be 100% explicit, then it would have been jdk.foreign.system.linker/jdk.foreign.native.linker - but what you picked is better than what it replaces).

**Pull request closed without merging**